doc/design/block.md

```diff
@@ -0,0 +1,173 @@
+
+Let's start from how a RNNOp is described using Block:
+
+```python
+v = some_op()
+m_boot = some_op()
+
+W = pd.Variable(shape=[20, 20])
+U = pd.Varable(shape=[20, 20])
+
+rnn = RNNOp()
+with rnn.stepnet() as net:
+  x = net.add_input(v)
+  h = net.add_memory(init=m_boot)
+  
+  fc_out = pd.matmul(W, x)
+  hidden_out = pd.matmul(U, h)
+  sum = pd.add_two(fc_out, hidden_out)
+  act = pd.sigmoid(sum)
+  # declare outputs
+  net.add_output(act, hidden_out)
+
+acts, hs = rnn()
+```
+
+This python program will be transformed into Protobuf messages which describe the model, a C++ Block class will create the corresponding Variables and Operators, and execute all the operators.
+
+## Block Implementation
+
+During the Protobuf message generation, the Block should store VarDesc (the Protobuf message which describes Variable) and OpDesc (the Protobuf message which describes Operator).
+
+VarDesc in a block should have its name scope to avoid local variables affect father block's name scope. 
+Child block's name scopes should inherit the father's so that OpDesc in child block can reference a VarDesc that stored in father block. For example
+
+```python
+a = pd.Varaible(shape=[20, 20])
+b = pd.fc(a, params=["fc.w", "fc.b"])
+
+rnn = pd.create_RNNOp()
+with rnn.stepnet() as net:
+    x = net.add_input(a)
+    # reuse fc's parameter
+    fc_without_b = pd.get_variable("fc.w")
+    net.add_output(fc_without_b)
+
+out = rnn()
+```
+the method `pd.get_variable` can help retrieve a Variable by a name, a Variable may store in a father block, but might be retrieved in a child block, so block should have a variable scope that supports inheritance.
+
+We can implement this idea in the following approach
+
+```c++
+class Block {
+public:
+  Block(Block* father) : father_block_{father} {}
+  // add a new VarDesc, return the pointer to enable other functions.
+  // NOTE Will check whether some variable called the same name.
+  VarDesc* AddVarDesc();
+  OpDesc* AddOpDesc();
+
+  // name: variable's name
+  // recursive: whether to find the variable in father's scope recursively.
+  VarDesc* FindVarDesc(const string& name, bool recursive=true);
+
+private:
+  Block* father_block_;
+  // descriptions
+  map<VarDesc> var_descs_;
+  map<OpDesc> op_descs_;
+}
+```
+
+after all the VarDescs and OpDescs are added into Block, the block is ready to run.
+During the running of Block, it first creates all Variables according to VarDescs and Operators according to OpDescs.
+
+What's more, each block will have its scope.
+
+The codes are as follows:
+
+```c++
+class Block : public OperatorBase {
+public:
+  // ...
+  void Run(const framework::Scope& scope,
+           const platform::DeviceContext& dev_ctx) const {
+    RuntimeInit();
+    LocalScopeInit(scope);
+
+    for (auto* op : ops_) {
+      op->Run(scope, dev_ctx);
+    }
+  }
+
+protected:
+  void LocalScopeInit(const framework::Scope &scope) {
```

Comment:
This function has only a few lines, and it only called by Run, should we move these a few lines into Run?

Reply:
yes.